servers/publikator/lib/github/index.js

```diff
@@ -132,22 +145,156 @@ module.exports = {
-      })
+
+    if (!rawCommit) {
+      return null
+    }
+    const commit = commitFromGQL(repo, rawCommit)
+    await redis.setAsync(redisKey, JSON.stringify(commit))
+    return commit
+  },
+  getCommits: async (repo, { first = 15, after, before }) => {
+    const { githubApolloFetch } = await createGithubClients()
+    const [login, repoName] = repo.id.split('/')
+    const {
+      data: {
+        repository: {
+          refs: {
+            nodes: heads
+          }
+        }
+      }
+    } = await githubApolloFetch({
+      query: `
+        query repository(
+          $login: String!,
+          $repoName: String!,
+          $maxRefs: Int
+          $maxCommits: Int,
+          $commitsSince: GitTimestamp,
+          $commitsUntil: GitTimestamp
+        ) {
+          repository(
+            owner: $login
+            name: $repoName
+          ) {
+            refs(refPrefix: "refs/heads/", first: $maxRefs) {
+              nodes {
+                name
+                target {
+                  ... on Commit {
+                    oid
+                    author {
+                      date
+                    },
+                    history(first: $maxCommits, until: $commitsUntil, since: $commitsSince) {
+                      pageInfo {
+                        hasNextPage
+                      }
+                      totalCount
+                      nodes {
+                        ... on Commit {
+                          oid
+                          author {
+                            email
+                            name
+                          }
+                          parents (first: 100){
+                            nodes {
+                              oid
+                            }
+                          }
+                          message
+                          committedDate
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      `,
+      variables: {
+        login,
+        repoName,
+        maxRefs: 100,
+        maxCommits: first,
+        commitsSince: before,
+        commitsUntil: after
+      }
+    })
+    const hasNextPage = heads.some(({target}) => target.history.pageInfo.hasNextPage)
+    const totalCount = heads.reduce((total, {target}) => total + target.history.totalCount, 0)
+
+    const commits = heads
+      .map(({ target }) =>
+        target.history.nodes
+          .map(
+            commit => commitFromGQL(repo, commit)
+          )
+      )
+      .reduce(
+        (acc, v) => acc.concat(v), []
+      )
+      .filter((v, i, arr) => arr.map(mapObj => mapObj.id).indexOf(v.id) === i)
+      .sort((a, b) => descending(a.date, b.date))
+      .slice(0, first)
+
+    return {
+      pageInfo: {
+        endCursor: (commits.length && commits.slice(-1)[0].date) || null,
+        startCursor: (commits.length && commits[0].date) || null,
```

Comment:
when querying with those cursor, will github include the first/last item again or will it be excluded? Does that mean the fundamental limitations is that when two commits share the same timestamp and occur around the pagination edge one could be missed?

Reply:
Both `since` and `until` are piped through to [git-log](https://git-scm.com/docs/git-log) somehow and both work by excluding the items with the exact equality on the timestamp, therefore it can even get used as cursor.  
As for the pagination corner case: It is not possible to have two commits added to the same repo at the same time in git. The same has got to be true for Github. The only scenario, where it has to handle possible timestamp equality is when syncing local and remote heads, which we never do. 
At least, that was my reasoning behind it.